libnmstate/ifaces/ethtool.py

```diff
@@ -71,28 +80,70 @@ def rx(self):
     def tx(self):
         return self._info.get(Ethtool.Pause.TX)
 
-    def pre_edit_validation_and_cleanup(self, original_desire):
+    def canonicalize(self, original_desire):
         """
         When AUTO_NEGOTIATION is enabled, RX and TX should be ignored.
         Log warnning if desired has AUTO_NEGOTIATION: True and RX/TX
         configured.
+        Remove RX/TX when AUTO_NEGOTIATION is enabled.
         """
         if self.autoneg and (
-            original_desire.rx is not None or original_desire.tx is not None
+            original_desire.get(Ethtool.Pause.RX) is not None
+            or original_desire.get(Ethtool.Pause.TX) is not None
         ):
             logging.warn(
                 "Ignoring RX/TX configure of ethtool PAUSE when "
                 "AUTO_NEGOTIATION enabled"
             )
-        self.canonicalize()
-
-    def canonicalize(self):
-        """
-        Remove RX/TX when AUTO_NEGOTIATION is enabled.
-        """
         if self.autoneg:
             self._info.pop(Ethtool.Pause.RX, None)
             self._info.pop(Ethtool.Pause.TX, None)
 
     def to_dict(self):
         return deepcopy(self._info)
+
+
+class IfaceEthtoolFeature:
+    _ETHTOOL_CLI_ALIASES = {
+        "rx": "rx-checksum",
+        "rx-checksumming": "rx-checksum",
+        "ufo": "tx-udp-fragmentation",
+        "gso": "tx-generic-segmentation",
+        "generic-segmentation-offload": "tx-generic-segmentation",
+        "gro": "rx-gro",
+        "generic-receive-offload": "rx-gro",
+        "lro": "rx-lro",
+        "large-receive-offload": "rx-lro",
+        "rxvlan": "rx-vlan-hw-parse",
+        "rx-vlan-offload": "rx-vlan-hw-parse",
+        "txvlan": "tx-vlan-hw-insert",
+        "tx-vlan-offload": "tx-vlan-hw-insert",
+        "ntuple": "rx-ntuple-filter",
+        "ntuple-filters": "rx-ntuple-filter",
+        "rxhash": "rx-hashing",
+        "receive-hashing": "rx-hashing",
+    }
+
+    def __init__(self, feature_info):
+        self._info = feature_info
+
+    def canonicalize(self, original_desire):
+        """
+        * Convert ethtool CLI alias to kernel names both in self and
+        original_desire
+        """
+        saved_keys = list(self._info.keys())
+        for key in saved_keys:
+            if key in IfaceEthtoolFeature._ETHTOOL_CLI_ALIASES:
+                kernel_key_name = IfaceEthtoolFeature._ETHTOOL_CLI_ALIASES[key]
+                self._info[kernel_key_name] = self._info[key]
+                original_desire[kernel_key_name] = self._info[key]
+                self._info.pop(key)
+                original_desire.pop(key, None)
+
+    def items(self):
+        for k, v in self._info.items():
+            yield k, v
```

Comment:
I do not like this name, it could be confusing with the dictionary function `items()`. But I do not have a better option.

Reply:
With this name, we don't even need a comment to explain what this function do. And it is actually doing the same thing as `items()` of dictionary does.